pkg/xds/generator/meshexternalservice/generator.go

```diff
@@ -0,0 +1,186 @@
+	"strings"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"github.com/kumahq/kuma/pkg/core/resources/apis/meshexternalservice/api/v1alpha1"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+	envoy_common "github.com/kumahq/kuma/pkg/xds/envoy"
+	envoy_clusters "github.com/kumahq/kuma/pkg/xds/envoy/clusters"
+	v3 "github.com/kumahq/kuma/pkg/xds/envoy/clusters/v3"
+	"github.com/kumahq/kuma/pkg/xds/envoy/listeners"
+	"github.com/kumahq/kuma/pkg/xds/envoy/names"
+	generator_core "github.com/kumahq/kuma/pkg/xds/generator/core"
+)
+
+// OriginMeshExternalService is a marker to indicate by which ProxyGenerator resources were generated.
+const OriginMeshExternalService = "meshexternalservice"
+
+var extensions map[string]Extension
+
+type Generator struct{}
+
+type Extension interface {
+	Generate(config *apiextensionsv1.JSON) (*core_xds.ResourceSet, error)
+}
+
+var _ generator_core.ResourceGenerator = Generator{}
+
+func RegisterExtension(name string, extension Extension) {
+	if extensions[name] != nil {
+		panic("extension " + name + " already registered")
+	}
+	extensions[name] = extension
+}
+
+// Generate generates MeshExternalService related resources
+func (g Generator) Generate(
+	ctx context.Context,
+	_ *core_xds.ResourceSet,
+	xdsCtx xds_context.Context,
+	proxy *core_xds.Proxy,
+) (*core_xds.ResourceSet, error) {
+	resources := core_xds.NewResourceSet()
+	meshExternalServices := xdsCtx.Mesh.Resources.MeshExternalServices().Items
+
+	for _, mes := range meshExternalServices {
+		res, err := g.generateResources(mes, proxy, xdsCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		resources.AddSet(res)
+	}
+
+	return resources, nil
+}
+
+func (g Generator) generateResources(mes *v1alpha1.MeshExternalServiceResource, proxy *core_xds.Proxy, xdsCtx xds_context.Context) (*core_xds.ResourceSet, error) {
+	// address not assigned, no need to generate resources
+	if mes.Status.VIP.IP == "" {
+		return nil, nil
+	}
+	resources := core_xds.NewResourceSet()
+
+	if mes.Spec.Extension != nil {
+		res, err := extensions[mes.Spec.Extension.Type].Generate(mes.Spec.Extension.Config) // pass tls / endpoints as well
+		if err != nil {
+			return nil, err
+		}
+		resources.AddSet(res)
+	} else {
+		cluster, err := createCluster(mes, proxy, xdsCtx)
+		if err != nil {
+			return nil, err
+		}
+		resources.Add(&core_xds.Resource{
+			Name:     cluster.GetName(),
+			Origin:   OriginMeshExternalService,
+			Resource: cluster,
+		})
+
+		listener, err := createListener(mes, proxy)
+		if err != nil {
+			return nil, err
+		}
+		if listener != nil {
+			resources.Add(&core_xds.Resource{
+				Name:     listener.GetName(),
+				Origin:   OriginMeshExternalService,
+				Resource: listener,
+			})
+		}
+	}
+
+	return resources, nil
+}
+
+func createCluster(mes *v1alpha1.MeshExternalServiceResource, proxy *core_xds.Proxy, xdsCtx xds_context.Context) (envoy_common.NamedResource, error) {
```

Comment:
How does this fit together with MeshHTTPRoute, which at the moment is responsible for listeners and clusters?

Reply:
Yes maybe that code should fall in MeshHTTPRoute and MeshTCPRoute instead which are sort of the Meta policies that create clusters (including subsets)